File: src/com/afollestad/polarupgradetool/MainBase.java
package com.afollestad.polarupgradetool;

import print.color.Ansi;
import print.color.ColoredPrinter;
import print.color.ColoredPrinterI;
import print.color.ColoredPrinterWIN;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;

/**
 * @author Aidan Follestad (afollestad)
 */
class MainBase {

    private final static String ARCHIVE_URL = "https://github.com/afollestad/polar-dashboard/archive/master.zip";
    public final static int BUFFER_SIZE = 2048;

    protected static File EXTRACTED_ZIP_ROOT;
    protected static File CURRENT_DIR;
    private final static String ARCHIVE_ROOT = File.separator + "polar-dashboard-master";

    public static String cleanupPath(String from) {
        if (from.startsWith(CURRENT_DIR.getAbsolutePath())) {
            from = from.substring(CURRENT_DIR.getAbsolutePath().length());
        } else if (from.startsWith(EXTRACTED_ZIP_ROOT.getAbsolutePath())) {
            from = from.substring(EXTRACTED_ZIP_ROOT.getAbsolutePath().length());
        }
        return from;
    }

    public static void LOG(String msg, Object... args) {
        if (args != null)
            msg = String.format(msg, args);
        Ansi.FColor color;
        if (msg.startsWith("[ERROR]")) {
            color = Ansi.FColor.RED;
        } else if (msg.startsWith("[INFO]") || msg.startsWith("[DETECTED]") || msg.startsWith("[RENAMING]")) {
            color = Ansi.FColor.CYAN;
        } else {
            color = Ansi.FColor.WHITE;
        }
        getPrinter(color, Ansi.BColor.BLACK).println(msg);
    }

    public static void PROGRESS(String label, long read, long total) {
        final int percent = (int) Math.ceil(((double) read / (double) total) * 100d);
        StringBuilder sb = new StringBuilder(13);
        sb.append('\r');
        if (label != null) {
            sb.append(label);
            sb.append("  ");
        }
        sb.append('[');
        final int numOfEqual = percent / 10;
        final int numOfSpace = 10 - numOfEqual;
        for (int i = 0; i < numOfEqual; i++) sb.append('=');
        for (int i = 0; i < numOfSpace; i++) sb.append(' ');
        sb.append("]");
        sb.append("   ");
        sb.append(Util.round(percent));
        sb.append("%  ");
        sb.append(Util.readableFileSizeMB(read));
        sb.append('/');
        sb.append(Util.readableFileSizeMB(total));
        System.out.print(sb.toString());
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    protected static boolean downloadArchive() {
        InputStream is = null;
        FileOutputStream os = null;
        try {
            URL url = new URL(ARCHIVE_URL);
            URLConnection conn = url.openConnection();
            is = conn.getInputStream();

            long contentLength;
            try {
                final String contentLengthStr = conn.getHeaderField("Content-Length");
                if (contentLengthStr == null || contentLengthStr.trim().isEmpty()) {
                    LOG("[ERROR]: No Content-Length header was returned by GitHub. Try running this app again.");
                    return false;
                }
                contentLength = Long.parseLong(contentLengthStr);
            } catch (Throwable e) {
                e.printStackTrace();
                LOG("[ERROR]: Failed to get the size of Polar's latest code archive. Please try running this app again.", e.getMessage());
                return false;
            }

            final File destZip = new File(CURRENT_DIR, "PolarLatest.zip");
            if (destZip.exists()) destZip.delete();
            os = new FileOutputStream(destZip);

            byte[] buffer = new byte[BUFFER_SIZE];
            int read;
            int totalRead = 0;

            LOG("[INFO]: Downloading a ZIP of Polar's latest code (%s)...", FileUtil.readableFileSize(contentLength));
            while ((read = is.read(buffer)) != -1) {
                os.write(buffer, 0, read);
                totalRead += read;
                PROGRESS(null, totalRead, contentLength);
            }

            PROGRESS(null, contentLength, contentLength);
            System.out.println();
            LOG("[INFO]: Download complete!");
            os.flush();

            Util.closeQuietely(is);
            Util.closeQuietely(os);

            EXTRACTED_ZIP_ROOT = new File(CURRENT_DIR, "PolarLatest");
            if (EXTRACTED_ZIP_ROOT.exists()) {
                LOG("[INFO]: Removed %d files/folders from %s.", FileUtil.wipe(EXTRACTED_ZIP_ROOT),
                        Main.cleanupPath(EXTRACTED_ZIP_ROOT.getAbsolutePath()));
            }

            LOG("[INFO]: Extracting %s to %s...", cleanupPath(destZip.getAbsolutePath()),
                    cleanupPath(EXTRACTED_ZIP_ROOT.getAbsolutePath()));
            UnzipUtil.unzip(destZip.getAbsolutePath(), EXTRACTED_ZIP_ROOT.getAbsolutePath());
            LOG("[INFO]: Extraction complete!\n");
            destZip.delete();
            EXTRACTED_ZIP_ROOT = new File(EXTRACTED_ZIP_ROOT, ARCHIVE_ROOT);
        } catch (Exception e) {
            LOG("[ERROR]: An error occurred during download or extraction: %s\n", e.getMessage());
            return false;
        } finally {
            Util.closeQuietely(is);
            Util.closeQuietely(os);
        }
        return true;
    }

    protected static ColoredPrinterI getPrinter(Ansi.FColor frontColor, Ansi.BColor backColor) {
        String os = System.getProperty("os.name");
        //System.out.println("DETECTED OS: " + os);
        if (os.toLowerCase().startsWith("win")) {
            return new ColoredPrinterWIN.Builder(1, false)
                    .foreground(frontColor).background(backColor).build();
        } else {
            return new ColoredPrinter.Builder(1, false)
                    .foreground(frontColor).background(backColor).build();
        }

    }
}

File: src/main/java/com/afollestad/polarupgradetool/MainBase.kt
package com.afollestad.polarupgradetool;

import com.afollestad.polarupgradetool.jfx.UICallback
import com.afollestad.polarupgradetool.utils.FileUtil
import com.afollestad.polarupgradetool.utils.UnzipUtil
import com.afollestad.polarupgradetool.utils.Util
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream
import java.net.URL

/**
 * @author Aidan Follestad (afollestad)
 */
open class MainBase {
    companion object {

        private val ARCHIVE_URL = "https://github.com/afollestad/polar-dashboard/archive/master.zip"
        private val ARCHIVE_ROOT = File.separator + "polar-dashboard-master"

        val BUFFER_SIZE = 2048
        var EXTRACTED_ZIP_ROOT: File? = null
        var CURRENT_DIR: File? = null

        fun cleanupPath(path: String): String {
            var from = path
            if (from.startsWith(CURRENT_DIR!!.absolutePath)) {
                from = from.substring(CURRENT_DIR!!.absolutePath.length)
            } else if (from.startsWith(EXTRACTED_ZIP_ROOT!!.absolutePath)) {
                from = from.substring(EXTRACTED_ZIP_ROOT!!.absolutePath.length)
            }
            return from
        }

        public fun LOG(msg: String) {
            println(msg)
        }

        fun PROGRESS(label: String?, read: Long, total: Long): String {
            val percent = Math.ceil(read.toDouble() / total.toDouble() * 100.0).toInt()
            val sb = StringBuilder(13)
            sb.append('\r')
            if (label != null) {
                sb.append(label)
                sb.append("  ")
            }
            sb.append('[')
            val numOfEqual = percent / 10
            val numOfSpace = 10 - numOfEqual
            for (i in 0..numOfEqual - 1) sb.append('=')
            for (i in 0..numOfSpace - 1) sb.append(' ')
            sb.append("]")
            sb.append("   ")
            sb.append(Util.round(percent.toFloat()))
            sb.append("%  ")
            sb.append(Util.readableFileSizeMB(read))
            sb.append('/')
            sb.append(Util.readableFileSizeMB(total))
            print(sb.toString())
            return "%s/%s (%s%%)".format(Util.readableFileSizeMB(read), Util.readableFileSizeMB(total), Util.round(percent.toFloat()))
        }

        var TRIES = 0

        @SuppressWarnings("ResultOfMethodCallIgnored")
        fun downloadArchive(uiCallback: UICallback): Boolean {
            var `is`: InputStream? = null
            var os: FileOutputStream? = null

            if (TRIES == 0) {
                LOG("[INFO]: Contacting GitHub...")
                uiCallback.onStatusUpdate("Contacting GitHub...")
            }

            try {
                val url = URL(ARCHIVE_URL)
                val conn = url.openConnection()
                `is` = conn.inputStream

                val contentLength: Long
                try {
                    val contentLengthStr = conn.getHeaderField("Content-Length")
                    if (contentLengthStr == null || contentLengthStr.trim { it <= ' ' }.isEmpty()) {
                        if (TRIES > 1) {
                            LOG("[ERROR]: No Content-Length header was returned by GitHub. Try running this app again.")
                            uiCallback.onErrorOccurred("GitHub did not report a Content-Length, please try again.")
                            return false
                        }
                        TRIES++
                        Thread.sleep(2000)
                        return downloadArchive(uiCallback)
                    }
                    contentLength = java.lang.Long.parseLong(contentLengthStr)
                } catch (e: Throwable) {
                    e.printStackTrace()
                    LOG("[ERROR]: Failed to get the size of Polar's latest code archive. Please try running this app again.")
                    uiCallback.onArchiveDownloadFailed("Failed to get the size of Polar's latest code archive. Please try running this app again.")
                    return false
                }

                val destZip = File(CURRENT_DIR, "PolarLatest.zip")
                if (destZip.exists()) destZip.delete()
                os = FileOutputStream(destZip)

                val buffer = ByteArray(BUFFER_SIZE)
                var read: Int
                var totalRead = 0

                LOG("[INFO]: Downloading a ZIP of Polar's latest code (${FileUtil.readableFileSize(contentLength)})...")
                uiCallback.onArchiveDownloadStarted(FileUtil.readableFileSize(contentLength))

                while (true) {
                    read = `is`!!.read(buffer);
                    if (read == -1) break;
                    os.write(buffer, 0, read)
                    totalRead += read
                    val progressStr = PROGRESS(null, totalRead.toLong(), contentLength)
                    uiCallback.onArchiveDownloadProgress(progressStr)
                }

                PROGRESS(null, contentLength, contentLength)
                println()
                LOG("[INFO]: Download complete!")
                uiCallback.onArchiveDownloadSuccess()
                os.flush()

                Util.closeQuietely(`is`)
                Util.closeQuietely(os)

                EXTRACTED_ZIP_ROOT = File(CURRENT_DIR, "PolarLatest")
                val destZipPath = destZip.absolutePath
                val extractedPath = EXTRACTED_ZIP_ROOT!!.absolutePath
                if (EXTRACTED_ZIP_ROOT!!.exists()) {
                    val removedCount = FileUtil.wipe(EXTRACTED_ZIP_ROOT!!)
                    LOG("[INFO]: Removed $removedCount files/folders from $extractedPath.")
                    uiCallback.onStatusUpdate("Removed $removedCount files/folders from $extractedPath.");
                }

                LOG("[INFO]: Extracting ${cleanupPath(destZipPath)} to ${cleanupPath(extractedPath)}...")
                uiCallback.onStatusUpdate("Extracting ${cleanupPath(destZipPath)} to ${cleanupPath(extractedPath)}...")
                UnzipUtil.unzip(destZipPath, extractedPath)
                LOG("[INFO]: Extraction complete!\n")
                uiCallback.onStatusUpdate("Extraction complete!")
                destZip.delete()
                EXTRACTED_ZIP_ROOT = File(EXTRACTED_ZIP_ROOT, ARCHIVE_ROOT)
            } catch (e: Exception) {
                LOG("[ERROR]: An error occurred during download or extraction: ${e.message}\n")
                uiCallback.onErrorOccurred("An error occurred during download or extraction: ${e.message}")
                return false
            } finally {
                Util.closeQuietely(`is`)
                Util.closeQuietely(os)
            }
            return true
        }
    }
}
